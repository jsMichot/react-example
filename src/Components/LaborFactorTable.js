import React, { Component } from 'react';
import _ from 'lodash';

export function reducer(state, action = {}) {
  switch (action.type) {
    case 'GetLaborFactorData':
      return [
        { id: 1, size: 1, class: '150', schedule: '', laborFactor: 1.2 },
        { id: 2, size: 1, class: '300', schedule: '', laborFactor: 2.2 },
        { id: 3, size: 2, class: '150', schedule: '', laborFactor: 1.2 },
        { id: 4, size: 2, class: '300', schedule: '', laborFactor: 2.2 },
        { id: 5, size: 3, class: '150', schedule: '', laborFactor: 1.2 },
        { id: 6, size: 3, class: '300', schedule: '', laborFactor: 1.3 },
        { id: 7, size: 3, class: '450', schedule: '', laborFactor: 1.4 },
        { id: 8, size: 3, class: '450', schedule: '80', laborFactor: 1.4 },
      ];
    default:
      return state;
  }
}

class LaborFactorTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      data: [],
      sizes: [],
      tempSizes: [],
      editing: false,
      addingColumn: false,
      addingRow: false,
      scheduleClassPairs: [],
      editValues: [],
      newRowId: 1,
      newRowClass: '',
      newRowSchedule: '',
    };
    this.getScheduleClassPairs = this.getScheduleClassPairs.bind(this);
    this.setLaborFactors = this.setLaborFactors.bind(this);
    this.cancelEdit = this.cancelEdit.bind(this);
    this.getSizes = this.getSizes.bind(this);
    this.editLaborFactors = this.editLaborFactors.bind(this);
    this.setLaborFactors = this.setLaborFactors.bind(this);
    this.addNewColumn = this.addNewColumn.bind(this);
    this.deleteNewColumn = this.deleteNewColumn.bind(this);
    this.saveNewColumn = this.saveNewColumn.bind(this);
    this.addNewRow = this.addNewRow.bind(this);
    this.saveNewRow = this.saveNewRow.bind(this);
    this.cancelNewRow = this.cancelNewRow.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.changeLaborFactor = this.changeLaborFactor.bind(this);
    this.changeTempSize = this.changeTempSize.bind(this);
  }

  componentDidMount() {
    const items = reducer(this.state, { type: 'GetLaborFactorData' });
    const updatedSizes = this.getSizes(items);
    const updatedScheduleClassPairs = this.getScheduleClassPairs(items, updatedSizes);
    this.setState((state) => (
      {
        data: items,
        scheduleClassPairs: updatedScheduleClassPairs,
        sizes: updatedSizes,
      }
    ), () => {
      console.dir(this.state);
    });
  }

  getScheduleClassPairs(data, sizes) {
    return data.reduce((uniquePairs, item) => {
      const currentPair = {
        schedule: item.schedule,
        class: item.class,
      };
      if (_.find(uniquePairs, currentPair)) {
        return uniquePairs;
      }
      sizes.forEach(size => {
        const classScheduleSizeMatch = _.find(data, {
          class: currentPair.class,
          schedule: currentPair.schedule,
          size,
        }) || { laborFactor: 0 };

        currentPair[size] = classScheduleSizeMatch.laborFactor;
      });
      return uniquePairs.concat(currentPair);
    }, []);
  }

  getSizes(data) {
    // debugger;
    const sizeMap = data.reduce(
      (acc, item) => ({ ...acc, [item.size]: true }),
      {}
    );
    return Object.keys(sizeMap).map(x => parseInt(x, 10));
  }

  editLaborFactors() {
    const newLaborFactorValues = this.state.scheduleClassPairs.map(pair =>
      Object.assign({}, pair));
    this.setState(state => (
      {
        editValues: newLaborFactorValues,
        editing: true,
      }
    ));
  }

  setLaborFactors() {
    this.setState(() => ({ editing: false }));
  }

  cancelEdit() {
    const previousScheduleClassPairs = this.state.editValues.map(pair =>
      Object.assign({}, pair));
    this.setState(state => ({
      scheduleClassPairs: previousScheduleClassPairs,
      editing: false,
    }));
  }

  addNewColumn() {
    const { sizes } = this.state;
    const temporarySizes = sizes.slice().concat((
      Math.floor(parseInt(sizes[sizes.length - 1], 10) + 1)
    ));
    this.setState(state => ({
      tempSizes: temporarySizes || [],
      addingColumn: true,
    }));
  }

  deleteNewColumn() {
    this.setState(() => ({ addingColumn: false }));
  }

  saveNewColumn() {
    const newSizes = _.uniq(this.state.tempSizes).sort();
    const newScheduleClassPairs = this.getScheduleClassPairs(this.state.data, newSizes);
    this.setState(() => ({
      sizes: newSizes,
      scheduleClassPairs: newScheduleClassPairs,
      addingColumn: false,
    }));
  }

  addNewRow() {
    this.setState(() => ({ addingRow: true }));
  }

  saveNewRow() {
    const { data, newRowId, newRowClass, newRowSchedule, sizes } = this.state;
    const newData = data.push({
      id: newRowId,
      class: newRowClass,
      schedule: newRowSchedule,
    });
    const newScheduleClassPairs = this.getScheduleClassPairs(newData, sizes);
    this.setState(() => ({
      data: newData,
      scheduleClassPairs: newScheduleClassPairs,
    }))
      .then(() => {
        this.cancelNewRow();
      });
  }

  cancelNewRow() {
    this.setState(() => ({
      newRowId: 1,
      newRowClass: '',
      newRowSchedule: '',
      addingRow: false,
    }));
  }

  handleChange(event) {
    console.log(event.target);
    const { name, value } = event.target;
    this.setState(() => ({ [name]: value }), () => {console.dir(this.state)});
  }

  changeLaborFactor(value, outerIndex, size) {
    console.log(value);
    // chore: improve form validation
    value = !value ? 0 : value;
    this.setState((state) => {
      state.scheduleClassPairs[outerIndex][size] = value; 
      return state;
    });
  }

  changeTempSize(value, index) {
    this.setState((state) => {
      state.tempSizes[index] = value;
      return state;
    });
  }

  render() {
    const { editing, addingColumn, addingRow, sizes, tempSizes, scheduleClassPairs } = this.state;

    return (
      <div>
        {!addingRow ?
          <div>
            <h1>Labor Factor Table</h1>
            {editing ? <button onClick={this.setLaborFactors}>Update Labor Factors</button> : null}
            {editing ? <button onClick={this.cancelEdit}>Cancel Edit</button> : null}
            {!addingColumn ? <button onClick={this.addNewColumn}>Add New Column</button> : null}
            {addingColumn ? <button onClick={this.deleteNewColumn}>Delete New Column</button> : null}
            {addingColumn ? <button onClick={this.saveNewColumn}>Save New Column</button> : null}
            {!addingRow ? <button onClick={this.addNewRow}>Add New Row</button> : null}
            <table>
              <thead>
                <tr>
                  <th className="tablePadding">Schedule</th>
                  <th className="tablePadding">Class</th>
                  {sizes.map(size => (
                    <th key={size}>
                      <span>
                        {size}
                      </span>
                    </th>
                  ))}
                  {addingColumn ? <th> <input type="text" defaultValue={tempSizes[tempSizes.length - 1]} onChange={event => {
                    const index = tempSizes.length - 1;
                    this.changeTempSize(event.target.value, index);
                  }} /> </th> : null}
                </tr>
              </thead>
              <tbody>
                {scheduleClassPairs.map((pair, outerIndex) => (
                  <tr onClick={this.editLaborFactor} key={outerIndex}>
                    <td>
                      {pair.schedule}
                    </td>
                    <td>
                      {pair.class}
                    </td>
                    {sizes.map((size, innerIndex) => (
                      <td key={innerIndex}>
                        {!editing ? <span onClick={this.editLaborFactors}>{scheduleClassPairs[outerIndex][size]}</span> :
                          <input type="text" defaultValue={scheduleClassPairs[outerIndex][size]} onChange={event => {this.changeLaborFactor(event.target.value, outerIndex, size)}} />
                        }
                      </td>
                    ))}
                  </tr>
                ))}
              </tbody>
            </table>
          </div>
          :
          <form>
            <h2>Add New Item</h2>
            <label>ID:
              <input type="text" name="newRowID" onChange={this.handleChange} />
            </label>
            <br />
            <label>CLASS:
              <input type="text" name="newRowClass" onChange={this.handleChange} />
            </label>
            <br />
            <label>SCHEDULE:
              <input type="text" name="newRowSchedule" onChange={this.handleChange} />
            </label>
            <br />
            <button onClick="this.cancelNewRow">Cancel New Row</button>
            <button onClick="this.saveNewRow">Save New Row</button>
          </form>
        }
      </div>
    );
  }
}

export default LaborFactorTable;
