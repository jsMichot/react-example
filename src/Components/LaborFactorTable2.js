import React, {Component} from 'react';

export function reducer(state, action = {}) {
  switch (action.type) {
    case 'GetLaborFactorTableData':
      return new Promise((resolve, reject) => {
        setTimeout(() => {
          resolve([
            {id: 1, size: 1, class: '150', schedule: '1', laborFactor: 1.2},
            {id: 2, size: 1, class: '300', schedule: '2', laborFactor: 2.2},
            {id: 3, size: 2, class: '150', schedule: '', laborFactor: 1.2},
            {id: 4, size: 2, class: '300', schedule: '2', laborFactor: 2.2},
            {id: 5, size: 3, class: '150', schedule: '', laborFactor: 1.2},
            {id: 6, size: 3, class: '300', schedule: '3', laborFactor: 1.3},
            {id: 7, size: 3, class: '450', schedule: '', laborFactor: 1.4},
          ]);
        }, 500);
      });
    case 'ComputeDataForLaborFactorTable':
      console.log(action.payload);
      const stateMap = action.payload.reduce(
        (acc, item) => {
          acc.sizes[item.size] = true;

          const key = `${item.class}|${item.schedule}`;
          acc.computedLookupTable[key] = acc.computedLookupTable[key] || {
            class: item.class,
            schedule: item.schedule,
          };
          acc.computedLookupTable[key][item.size] = item;
          return acc;
        },
        {sizes: {}, computedLookupTable: {}}
      );
      stateMap.sizes = Object.keys(stateMap.sizes).map(x => parseInt(x, 10));
      stateMap.rows = Object.keys(stateMap.computedLookupTable).map(
        x => stateMap.computedLookupTable[x]
      );
      console.log(stateMap);
      return {
        ...state,
        sizes: stateMap.sizes,
        computedLookupTable: stateMap.rows,
      };
    default:
      return state;
  }
}

class LaborFactorTable2 extends Component {
  state = {};

  componentDidMount() {
    reducer(this.state, {type: 'GetLaborFactorTable'}).then(data => {
      reducer(this.state, {
        type: 'ComputeDataForLaberFactorTable',
        payload: data,
      });
    });
  }

  render() {
    return <div />;
  }
}

export default LaborFactorTable2;
