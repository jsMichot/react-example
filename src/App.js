import React, {Component} from 'react';
import _ from 'lodash';
import LaborFactorTable from './Components/LaborFactorTable';

class App extends Component {
  state = {
    count: 1,
  };
  render() {
    return (
      <div>
        <LaborFactorTable />
      </div>
    );
  }
}

export default App;
