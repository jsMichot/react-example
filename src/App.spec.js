import React from 'react';
import {create} from 'react-test-renderer';
import {reducer} from './Components/LaborFactorTable';
import App from '../App';

xit('should render component and its current state', () => {
  const component = create(<App />);
  const initialState = {count: 3};
  component.root.instance.setState(initialState);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

xit('component should show increment size if provided', () => {
  const component = create(<App step={5} />);
  const initialState = {count: 3};
  component.root.instance.setState(initialState);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe('LaborFactorTable', () => {
  beforeEach(() => {
    create(<App />);
  });
  
  it('action: GetLaborFactorData should return the proper items', () => {
  // arrange
    const currentState = {data: []};
    const action = {type: 'GetLaborFactorData'};
    // act
    const nextState = reducer(currentState, action);
    // assert
    expect(nextState.data).toEqual([
      {id: 1, size: 1, class: '150', schedule: '', laborFactor: 1.2},
      {id: 1, size: 1, class: '300', schedule: '', laborFactor: 2.2},
      {id: 1, size: 2, class: '150', schedule: '', laborFactor: 1.2},
      {id: 1, size: 2, class: '300', schedule: '', laborFactor: 2.2},
      {id: 1, size: 3, class: '150', schedule: '', laborFactor: 1.2},
      {id: 1, size: 3, class: '300', schedule: '', laborFactor: 1.3},
      {id: 1, size: 3, class: '450', schedule: '', laborFactor: 1.4},
      {id: 1, size: 3, class: '450', schedule: '80', laborFactor: 1.4},
    ]);
  });

  it('editLaborFactor should set editValues to scheduleClassPairs', () => {
    this.scheduleClassPairs = [{ foo: 1 }];
    this.editLaborFactor();

    expect(this.editing).toBe(true);
    expect(this.editValues).toEqual([{ foo: 1 }]);
  });

  it('setLaborFactor should set scheduleClassPairs to editValues', () => {
    this.editing = true;
    this.setLaborFactors();

    expect(this.editing).toBe(false);
  });

  it('cancelEdit should set scheduleClassPairs to editValues and editing to false', () => {
    this.editValues = [{ foo: 1 }];
    this.scheduleClassPairs = [{ foo: 2 }];
    this.cancelEdit();
    expect(this.scheduleClassPairs).toEqual([{ foo: 1 }]);
    expect(this.editing).toBe(false);
  });

  it('when editLaborFactor is called editValues and scheduleClassPairs should not point to the same array nor the same objects', () => {
    this.scheduleClassPairs = [{ foo: 1 }];
    this.editLaborFactor();
    this.editValues[0].foo = 2;

    expect(this.scheduleClassPairs).toEqual([{ foo: 1 }]);
  });

  it('when cancelEdit is called editValues and scheduleClassPairs should not point to the same array nor the same objects (i.e. editValues retains the pre-editing values of scheduleClassPairs)', () => {
    this.scheduleClassPairs = [{ foo: 1 }];
    this.editLaborFactor();
    this.scheduleClassPairs[0].foo = 2;
    this.cancelEdit();

    expect(this.editValues).toEqual([{ foo: 1 }]);
  });

  it('when addNewColumn is called sizes should be cloned to tempSizes (separate arrays) with an added size of the next whole number', () => {
    this.sizes = [1, 2, 3];
    this.addNewColumn();
    expect(this.tempSizes).toEqual([1, 2, 3, 4]);
    this.sizes = [1, 2, 3, 4.34];
    this.addNewColumn();
    expect(this.tempSizes).toEqual([1, 2, 3, 4.34, 5]);
  });

  it('when addNewColumn is called addingColumn should be set to true', () => {
    this.addNewColumn();
    expect(this.addingColumn).toBe(true);
  });
  it('when saveNewColumn is called sizes should not contain any duplicates', () => {
    this.sizes = [1, 2, 3, 3];
    this.tempSizes = [1, 2, 3, 3, 3];
    this.saveNewColumn();
    expect(this.sizes).toEqual([1, 2, 3]);
  });

  it('when deleteNewColumn is called addingColumn should be set to false', () => {
    this.deleteNewColumn();
    expect(this.addingColumn).toBe(false);
  });

  it('when saveNewColumn is called sizes should be sorted', () => {
    const data = [
      { id: 1, size: 1, class: '150', schedule: '1', laborFactor: 1.2 },
      { id: 1, size: 1, class: '300', schedule: '2', laborFactor: 2.2 },
      { id: 1, size: 2, class: '150', schedule: '', laborFactor: 1.2 },
      { id: 1, size: 2, class: '300', schedule: '2', laborFactor: 2.2 },
      { id: 1, size: 3, class: '150', schedule: '', laborFactor: 1.2 },
      { id: 1, size: 3, class: '300', schedule: '3', laborFactor: 1.3 },
      { id: 1, size: 3, class: '450', schedule: '', laborFactor: 1.4 },
    ];
    this.sizes = [1, 2, 3];
    this.addNewColumn();
    this.sizes.push(2.5);
    this.saveNewColumn();
    expect(this.sizes).toEqual([1, 2, 2.5, 3, 4]);
  });

  it('addNewRow should set addingRow to true', () => {
    this.addingRow = false;
    this.addNewRow();
    expect(this.addingRow).toBe(true);
  });

  it('saveNewRow should add the new item to data', () => {
    this.data = [
      { id: 1, size: 1, class: '150', schedule: '1', laborFactor: 1.2 },
      { id: 1, size: 1, class: '300', schedule: '2', laborFactor: 2.2 },
      { id: 1, size: 2, class: '150', schedule: '', laborFactor: 1.2 },
      { id: 1, size: 2, class: '300', schedule: '2', laborFactor: 2.2 },
      { id: 1, size: 3, class: '150', schedule: '', laborFactor: 1.2 },
      { id: 1, size: 3, class: '300', schedule: '3', laborFactor: 1.3 },
      { id: 1, size: 3, class: '450', schedule: '', laborFactor: 1.4 },
    ];
    this.newRowId = 1;
    this.newRowClass = '200';
    this.newRowSchedule = '';
    this.saveNewRow();
    expect(this.data).toEqual([
      { id: 1, size: 1, class: '150', schedule: '1', laborFactor: 1.2 },
      { id: 1, size: 1, class: '300', schedule: '2', laborFactor: 2.2 },
      { id: 1, size: 2, class: '150', schedule: '', laborFactor: 1.2 },
      { id: 1, size: 2, class: '300', schedule: '2', laborFactor: 2.2 },
      { id: 1, size: 3, class: '150', schedule: '', laborFactor: 1.2 },
      { id: 1, size: 3, class: '300', schedule: '3', laborFactor: 1.3 },
      { id: 1, size: 3, class: '450', schedule: '', laborFactor: 1.4 },
      { id: 1, class: '200', schedule: '' },
    ]);
  });

  it('cancelNewRow should set newRowId to 1', () => {
    this.newRowId = 4;
    this.cancelNewRow();
    expect(this.newRowId).toBe(1);
  });

  it('cancelNewRow should set newRowClass to an empty string', () => {
    this.newRowClass = '350';
    this.cancelNewRow();
    expect(this.newRowClass).toBe('');
  });

  it('cancelNewRow should set newRowSchedule to an empty string', () => {
    this.newRowSchedule = '3';
    this.cancelNewRow();
    expect(this.newRowSchedule).toBe('');
  });

  it('cancelNewRow should set addingRow to false', () => {
    this.addingRow = true;
    this.cancelNewRow();
    expect(this.addingRow).toBe(false);
  });

});
