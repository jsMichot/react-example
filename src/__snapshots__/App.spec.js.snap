// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render: cell click triggers editing & autofocus 1`] = `
Array [
  <table>
    <thead>
        <tr>
            <th>
                Class
            </th>
            <th>
                Schedule
            </th>
            <th>
                1
            </th>
            <th>
                2
            </th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td
                onClick={[Function]}
            >
                150
            </td>
            <td
                onClick={[Function]}
            >
                80
            </td>
            <td
                onClick={[Function]}
            >
                1.2
            </td>
            <td
                onClick={[Function]}
            >
                0
            </td>
        </tr>
        <tr>
            <td
                onClick={[Function]}
            >
                300
            </td>
            <td
                onClick={[Function]}
            >
                
            </td>
            <td
                onClick={[Function]}
            >
                <input
                    autoFocus={false}
                    defaultValue={2.2}
                    name="laborFactor"
                    onBlur={[Function]}
                />
            </td>
            <td
                onClick={[Function]}
            >
                <input
                    autoFocus={true}
                    defaultValue={0}
                    name="laborFactor"
                    onBlur={[Function]}
                />
            </td>
        </tr>
        <tr>
            <td
                onClick={[Function]}
            >
                150
            </td>
            <td
                onClick={[Function]}
            >
                
            </td>
            <td
                onClick={[Function]}
            >
                0
            </td>
            <td
                onClick={[Function]}
            >
                1.2
            </td>
        </tr>
    </tbody>
</table>,
  <div>
    Mutations:
    <strong>
        Adds =
         
        0
    </strong>
     | 
    <strong>
        Updates =
         
        0
    </strong>
</div>,
]
`;

exports[`render: default with cell autofocus 1`] = `
Array [
  <table>
    <thead>
        <tr>
            <th>
                Class
            </th>
            <th>
                Schedule
            </th>
            <th>
                1
            </th>
            <th>
                2
            </th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td
                onClick={[Function]}
            >
                150
            </td>
            <td
                onClick={[Function]}
            >
                80
            </td>
            <td
                onClick={[Function]}
            >
                <input
                    autoFocus={false}
                    defaultValue={1.2}
                    name="laborFactor"
                    onBlur={[Function]}
                />
            </td>
            <td
                onClick={[Function]}
            >
                <input
                    autoFocus={true}
                    defaultValue={0}
                    name="laborFactor"
                    onBlur={[Function]}
                />
            </td>
        </tr>
        <tr>
            <td
                onClick={[Function]}
            >
                300
            </td>
            <td
                onClick={[Function]}
            >
                
            </td>
            <td
                onClick={[Function]}
            >
                2.2
            </td>
            <td
                onClick={[Function]}
            >
                0
            </td>
        </tr>
        <tr>
            <td
                onClick={[Function]}
            >
                150
            </td>
            <td
                onClick={[Function]}
            >
                
            </td>
            <td
                onClick={[Function]}
            >
                0
            </td>
            <td
                onClick={[Function]}
            >
                1.2
            </td>
        </tr>
    </tbody>
</table>,
  <div>
    Mutations:
    <strong>
        Adds =
         
        0
    </strong>
     | 
    <strong>
        Updates =
         
        0
    </strong>
</div>,
]
`;

exports[`render: default with data 1`] = `
Array [
  <table>
    <thead>
        <tr>
            <th>
                Class
            </th>
            <th>
                Schedule
            </th>
            <th>
                1
            </th>
            <th>
                2
            </th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td
                onClick={[Function]}
            >
                150
            </td>
            <td
                onClick={[Function]}
            >
                80
            </td>
            <td
                onClick={[Function]}
            >
                1.2
            </td>
            <td
                onClick={[Function]}
            >
                0
            </td>
        </tr>
        <tr>
            <td
                onClick={[Function]}
            >
                300
            </td>
            <td
                onClick={[Function]}
            >
                
            </td>
            <td
                onClick={[Function]}
            >
                2.2
            </td>
            <td
                onClick={[Function]}
            >
                0
            </td>
        </tr>
        <tr>
            <td
                onClick={[Function]}
            >
                150
            </td>
            <td
                onClick={[Function]}
            >
                
            </td>
            <td
                onClick={[Function]}
            >
                0
            </td>
            <td
                onClick={[Function]}
            >
                1.2
            </td>
        </tr>
    </tbody>
</table>,
  <div>
    Mutations:
    <strong>
        Adds =
         
        0
    </strong>
     | 
    <strong>
        Updates =
         
        0
    </strong>
</div>,
]
`;

exports[`render: default with edited row 1`] = `
Array [
  <button
    onClick={[Function]}
>
    <em>
        Add Size
    </em>
</button>,
  <br />,
  <table>
    <thead>
        <tr>
            <th>
                Class
            </th>
            <th>
                Schedule
            </th>
            <th>
                1
            </th>
            <th>
                2
            </th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td
                onClick={[Function]}
            >
                150
            </td>
            <td
                onClick={[Function]}
            >
                80
            </td>
            <td
                onClick={[Function]}
            >
                <input
                    autoFocus={true}
                    defaultValue={1.2}
                    name="laborFactor"
                    onBlur={[Function]}
                />
            </td>
            <td
                onClick={[Function]}
            >
                <input
                    autoFocus={false}
                    defaultValue={0}
                    name="laborFactor"
                    onBlur={[Function]}
                />
            </td>
        </tr>
        <tr>
            <td
                onClick={[Function]}
            >
                300
            </td>
            <td
                onClick={[Function]}
            >
                
            </td>
            <td
                onClick={[Function]}
            >
                2.2
            </td>
            <td
                onClick={[Function]}
            >
                0
            </td>
        </tr>
        <tr>
            <td
                onClick={[Function]}
            >
                150
            </td>
            <td
                onClick={[Function]}
            >
                
            </td>
            <td
                onClick={[Function]}
            >
                0
            </td>
            <td
                onClick={[Function]}
            >
                1.2
            </td>
        </tr>
    </tbody>
</table>,
  <div>
    Mutations:
    <strong>
        Adds =
         
        0
    </strong>
     | 
    <strong>
        Updates =
         
        0
    </strong>
</div>,
]
`;
